packages/loader/container-loader/src/virtualDeltaManager.ts

```diff
@@ -0,0 +1,278 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { TypedEventEmitter, assert } from "@fluidframework/common-utils";
+import { IConnectionDetails, IDeltaHandlerStrategy, IDeltaQueue, IDeltaQueueEvents } from "@fluidframework/container-definitions";
+import { IDocumentService } from "@fluidframework/driver-definitions";
+import { IDocumentMessage, ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { IConnectionManager, IConnectionManagerFactoryArgs } from "./contracts";
+import { DeltaManager } from "./deltaManager";
+
+const GroupedBatchOpType = "groupedBatch";
+
+class VirtualDeltaQueueProxy<T> extends TypedEventEmitter<IDeltaQueueEvents<T>> implements IDeltaQueue<T> {
+    private readonly _trackedListeners: Map<string, (...args: any[]) => void> = new Map();
+
+    constructor(
+        private readonly _queue: IDeltaQueue<T>,
+        private readonly _virtualizeValue: (value: T, storeValue?: boolean) => T,
+    ) {
+        super();
+
+        this.on("newListener", (event, _listener) => {
+            assert((this.listeners(event).length !== 0) === this._trackedListeners.has(event), "mismatch");
+            if (!this._trackedListeners.has(event)) {
+                let listener: (...args: any[]) => void;
+
+                // eslint-disable-next-line unicorn/prefer-ternary
+                if (event === "push" || event === "op") {
+                    listener = (task: T) => { this.emit(event, this._virtualizeValue(task, true /* storeValue */)); }
+                } else {
+                    listener = (...args: any[]) => { this.emit(event, ...args); };
+                }
+
+                this.addTrackedListener(event, listener);
+            }
+        });
+    }
+
+    private addTrackedListener(event: string, listener: (...args: any[]) => void) {
+        (this._queue as any).on(event, listener); // TODO: find better cast than "any"
+        assert(!this._trackedListeners.has(event), "double tracked listener");
+        this._trackedListeners.set(event, listener);
+    }
+
+    private removeTrackedListeners() {
+        for (const [event, listener] of this._trackedListeners.entries()) {
+            (this._queue as any).off(event, listener); // TODO: find better cast than "any"
+        }
+        this._trackedListeners.clear();
+    }
+
+    public peek(): T | undefined {
+        const value = this._queue.peek();
+        return value !== undefined ? this._virtualizeValue(value) : undefined;
+    }
+
+    public toArray(): T[] {
```

Comment:
Is this method used? can we remove it / deprecate it?

Reply:
This is being used by the ScheduleManager ctor